lib/autoupdate/tools/updater.go

```diff
@@ -0,0 +1,391 @@
+	// to remain available after downloading archives.
+	reservedFreeDisk = 10 * 1024 * 1024 // 10 Mb
+	// lockFileName is file used for locking update process in parallel.
+	lockFileName = ".lock"
+	// updatePackageSuffix is directory suffix used for package extraction in tools directory.
+	updatePackageSuffix = "-update-pkg"
+)
+
+var (
+	// // pattern is template for response on version command for client tools {tsh, tctl}.
+	pattern = regexp.MustCompile(`(?m)Teleport v(.*) git`)
+)
+
+// Option applies an option value for the Updater.
+type Option func(u *Updater)
+
+// WithBaseURL defines custom base url for the updater.
+func WithBaseURL(baseURL string) Option {
+	return func(u *Updater) {
+		u.baseURL = baseURL
+	}
+}
+
+// WithClient defines custom http client for the Updater.
+func WithClient(client *http.Client) Option {
+	return func(u *Updater) {
+		u.client = client
+	}
+}
+
+// Updater is updater implementation for the client tools auto updates.
+type Updater struct {
+	toolsDir     string
+	localVersion string
+	tools        []string
+
+	baseURL string
+	client  *http.Client
+}
+
+// NewUpdater initializes the updater for client tools auto updates. We need to specify the list
+// of tools (e.g., `tsh`, `tctl`) that should be updated, the tools directory path where we
+// download, extract package archives with the new version, and replace symlinks (e.g., `$TELEPORT_HOME/bin`).
+// The base URL of the CDN with Teleport packages and the `http.Client` can be customized via options.
+func NewUpdater(tools []string, toolsDir string, localVersion string, options ...Option) *Updater {
+	updater := &Updater{
+		tools:        tools,
+		toolsDir:     toolsDir,
+		localVersion: localVersion,
+		baseURL:      baseURL,
+		client:       http.DefaultClient,
+	}
+	for _, option := range options {
+		option(updater)
+	}
+
+	return updater
+}
+
+// CheckLocal is run at client tool startup and will only perform local checks.
+func (u *Updater) CheckLocal() (string, bool) {
+	// Check if the user has requested a specific version of client tools.
+	requestedVersion := os.Getenv(teleportToolsVersionEnv)
+	switch {
+	// The user has turned off any form of automatic updates.
+	case requestedVersion == "off":
+		return "", false
+	// Requested version already the same as client version.
+	case u.localVersion == requestedVersion:
+		return requestedVersion, false
+	}
+
+	// If a version of client tools has already been downloaded to
+	// tools directory, return that.
+	toolsVersion, err := checkToolVersion(u.toolsDir)
+	if err != nil {
+		return "", false
+	}
+	// The user has requested a specific version of client tools.
+	if requestedVersion != "" && requestedVersion != toolsVersion {
+		return requestedVersion, true
+	}
+
+	return toolsVersion, false
+}
+
+// CheckRemote will check against Proxy Service if client tools need to be
+// updated.
+func (u *Updater) CheckRemote(ctx context.Context, proxyAddr string) (string, bool, error) {
```

Comment:
Document return params

Reply:
Still not clear to me what the bool means by reading the doc. Could you explain that in the godoc?
Might also help to name the params:
```
(version string, changed bool, er error)
```